### DISCUSSION TITLE
State of Full Text Fuzzy Search February 2022

### DISCUSSION THREAD

**Discussion by hjaber** [supabase/supabase#5435] [Questions] [answered]
While trying to create a front end search bar, have been really impressed how simple and great `textSearch` with `websearch` work combined with a `gin`.

I'm not a postgres expert so trying to optimize for  performance is difficult for me to grasp, but it appears that the `fuzzystrmatch` and `pg_trgm` extensions and different indexing types could be the best option. What is the best way today or near future to implement a frontend fuzzy search? 

From what I can tell, it appears to be creating a custom rpc function utilizing either `fuzzystrmatch` or `pg_trgm`.

I understand there are limitations to postgres' searching, but following this [guide](https://supabase.com/docs/guides/database/full-text-search), the search term `goble` will not find `goblet` or `rabit` finding `rabbit`.

I know I am asking for a lot, but my perfect solution would be fuzzy full text search with highlighting. I am trying to create a search bar that would search posts containing common english words and proper nouns. This would eliminate my need for Algolia or elasticsearch.

Much appreciated
```
    const { data, error } = await supabase
      .from("books")
      .select()
      .textSearch("fts", searchTerm, {
        type: "websearch",
        config: "english",
      });
```

https://user-images.githubusercontent.com/41503068/153719992-3ec23e5c-c988-4fd7-a2b7-53c7a2312f88.mov

**Comment by silentworks** [comment, marked as answer]
You can install any of those extensions in your Supabase instance and create a `.rpc` to execute the query. You can create a function inside of the SQL editor in the Supabase UI, here is an example using the table name `books` with the `pg_trgm` extension enabled.

Enable the extension by using the UI or from the SQL Editor
```sql
CREATE EXTENSION pg_trgm;
```

Postgres function
```sql
CREATE OR REPLACE FUNCTION public.search_books(book_term varchar)
RETURNS table(id int, title text, author text)
LANGUAGE plpgsql
AS $$
	begin
		return query
			SELECT b.id, b.title, b.author 
			FROM books b 
			WHERE book_term % ANY(STRING_TO_ARRAY(b.title, ' '));
	end;
$$;
```
You can then test this inside of the SQL editor using a SELECT query
```sql
SELECT * FROM search_books('Poter');
```
And from the supabase-js library you can do
```js
const { data, error } = supabase
  .rpc('search_books', { book_term: 'Poter' })
```
I hope this helps.

**Reply by asciidiego** [reply]
That was very helpful. Thank you sir.

**Reply by hdoshi2** [reply]
This was really helpful, do you by any chance know how to modify this `CREATE EXTENSION pg_trgm;` so that it is applied to the public tables. It can be done through the UI but would love to do it in SQL.

**Reply by hdoshi2** [reply]
Ok figured it out: `CREATE EXTENSION pg_trgm SCHEMA public;`

**Reply by mirsahib** [reply]
Thank you for such a details answer but the postgres function don't work for me,I am getting `Failed to run sql query: structure of query does not match function result type` but i figured a working solution
```sql
CREATE OR REPLACE FUNCTION search_products(product_title text)
RETURNS SETOF products
AS $$
BEGIN
    RETURN QUERY
        SELECT *
        FROM products p 
        WHERE product_title % ANY(STRING_TO_ARRAY(p.title, ' '));
END;
$$ LANGUAGE plpgsql; 
```
Changed i made is instead of returning a table its returning setof row of that table

**Reply by softmarshmallow** [reply]
No way to do this without functions ?

**Comment by hjaber** [comment]
Thank you @silentworks, this is very helpful for understanding how to create a custom search utilizing postgres functions and supabase-js.

I need to learn more about the nuances of `textSearch`, `pg-trgm`, & `fuzzystrmatch` and test the different solutions utilizing indexes to see what is the best solution in terms of performance/relevancy.

**Reply by silentworks** [reply]
I haven't personally run any benchmark on any of the solutions available, but I found this blog post which has some benchmarks in it. https://mazeez.dev/posts/pg-trgm-similarity-search-and-fast-like

This is also a good read and has some really good examples that show how you can save the indexes into a materialized view to optimise the querying. http://rachbelaid.com/postgres-full-text-search-is-good-enough/

**Reply by hjaber** [reply]
> I haven't personally run any benchmark on any of the solutions available, but I found this blog post which has some benchmarks in it. https://mazeez.dev/posts/pg-trgm-similarity-search-and-fast-like
> 
> This is also a good read and has some really good examples that show how you can save the indexes into a materialized view to optimise the querying. http://rachbelaid.com/postgres-full-text-search-is-good-enough/

Thanks for sharing, both articles are really helpful. Postgres full text search is powerful once you have it dialed in.

Without any real optimization beyond `gin` indexing, I am seeing impressive results on a 150,000 row dataset for  `textSearch`, `pg_trgm`, and `ilike`

`textSearch: 0.135ms`
`pg_trgm similiarity: 15ms`
`ilike: 0.3ms`

`pg_trgm` is probably what I'm looking for to provide a fast and easy fuzzy search. I am currently using `textSearch` with  prefix matching `:*` which I've been happy with but want the fuzzy functionality.

**Reply by zidkim** [reply]
@silentworks Thank you for the really clear explanation. Sorry to ask this, but could you please take this a step further and show how to implement a GIN index? I've tried...

`create index books_fts on books using gin (fts);`

...followed by your rpc function while swapping the `b.title` to b.fts `WHERE book_term % ANY(STRING_TO_ARRAY(b.title, ' '));` but can't seem to get it to work.

**Reply by RomanRC99** [reply]
Anyone solved the indexes issue for pg_trgm? I get this errors: 

```
2025-07-25 11:02:09 [ INFO] Executing: CREATE INDEX IF NOT EXISTS idx_units_street_gin ON public.units USING gin (street)
2025-07-25 11:02:09 [ERROR] data type text has no default operator class for access method "gin"
2025-07-25 11:03:50 [ INFO] Executing: CREATE INDEX IF NOT EXISTS idx_units_street_gin ON public.units USING gin (street gin_trgm_ops)
2025-07-25 11:03:50 [ERROR] operator class "gin_trgm_ops" does not exist for access method "gin"

```

**Comment by ghost** [comment]
For those who still struggles:

https://dba.stackexchange.com/a/103823

I tried just this. For me it worked better than above answer
```sql
select *
from   users
where  name % 'query'
order  by (name <-> 'query')
limit  10;
```

or for suggestion like solution

```sql
select * from users 
where name like ('%' || 'query' || '%') or  name like '%'
order by name <-> 'query'
limit 10
```

these solutions are good for dictinaries where words are short, also no need to index

**Comment by organicnz** [comment]
Thanks a lot for the great post and comments guys. By any chance, how does one go about making a cURL REST API call? Already tried this call `curl -X GET -H "apikey: $supabase_anon_key" "https://iazmjxxxxxxxxx.supabase.co/rest/v1/posts?select=*&filter=post_name.ilike.%seattle%"` but it returns an HTML spaghetti hehe :))

I'd like to create an `API Calls` in FlutterFlow with customised variables from this cURL call :)

```
-- Drop the existing index and extension (if they exist)
DROP INDEX IF EXISTS idx_posts_trgm;
DROP EXTENSION IF EXISTS pg_trgm;

-- Create the pg_trgm extension
CREATE EXTENSION pg_trgm;

-- Create a GIN index on the relevant columns
CREATE INDEX idx_posts_trgm ON posts USING gin (
  post_name gin_trgm_ops,
  post_address gin_trgm_ops,
  post_description gin_trgm_ops,
  network gin_trgm_ops,
  transportation gin_trgm_ops,
  website gin_trgm_ops,
  linktree gin_trgm_ops,
  donation gin_trgm_ops,
  donation_rules gin_trgm_ops,
  location_type gin_trgm_ops,
  local_map gin_trgm_ops,
  news_articles gin_trgm_ops,
  label gin_trgm_ops,
  food_status gin_trgm_ops,
  prefill_form gin_trgm_ops,
  post_type gin_trgm_ops
);

-- Set similarity threshold (once per session)
SET pg_trgm.similarity_threshold = 0.6;

-- Perform the search query
SELECT *
FROM posts
WHERE post_name LIKE ('%' || 'seattle' || '%')
   OR post_address LIKE ('%' || 'seattle' || '%')
   OR post_description LIKE ('%' || 'seattle' || '%')
   OR network LIKE ('%' || 'seattle' || '%')
   OR transportation LIKE ('%' || 'seattle' || '%')
   OR website LIKE ('%' || 'seattle' || '%')
   OR linktree LIKE ('%' || 'seattle' || '%')
   OR donation LIKE ('%' || 'seattle' || '%')
   OR donation_rules LIKE ('%' || 'seattle' || '%')
   OR location_type LIKE ('%' || 'seattle' || '%')
   OR local_map LIKE ('%' || 'seattle' || '%')
   OR news_articles LIKE ('%' || 'seattle' || '%')
   OR label LIKE ('%' || 'seattle' || '%')
   OR food_status LIKE ('%' || 'seattle' || '%')
   OR prefill_form LIKE ('%' || 'seattle' || '%')
   OR post_type LIKE ('%' || 'seattle' || '%')
ORDER BY (post_name <-> 'seattle'), (post_address <-> 'seattle'), created_at DESC
LIMIT 50;
```
